src/job.rs

```diff
@@ -1,16 +1,102 @@
+use anyhow::{bail, Context, Result};
 use std::collections::HashMap;
-use std::io;
+use std::fs;
+use std::path::{Path, PathBuf};
 use std::process::{Command, Output};
+use tempfile::Builder;
+use walkdir::WalkDir;
 
 #[derive(Debug, PartialEq, Eq)]
 pub struct Job {
     pub command: String,
     pub arguments: Vec<String>,
     pub environment: HashMap<String, String>,
+    pub working_directory: PathBuf,
+    pub inputs: Vec<PathBuf>,
+    pub outputs: Vec<PathBuf>,
 }
 
 impl Job {
-    pub fn run(&self) -> io::Result<Output> {
+    pub fn run(&self) -> Result<Output> {
+        let work_dir = Builder::new()
+            .prefix(format!("job-{}", self.command).as_str())
```

Comment:
In the future I think we'll want to use `bumpalo` for this project, and [`bumpalo::format!`](https://docs.rs/bumpalo/3.7.0/bumpalo/macro.format.html) here, but not necessary for this PR!